util/codec/codec.go

```diff
@@ -352,6 +352,240 @@ func encodeHashChunkRowIdx(sc *stmtctx.StatementContext, row chunk.Row, tp *type
+				b = (*[unsafe.Sizeof(f)]byte)(unsafe.Pointer(&f))[:]
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeDouble:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				row = chk.GetRow(i)
+				buf[0] = floatFlag
+				b = row.GetRaw(colIdx)
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = compactBytesFlag
+				b = row.GetBytes(colIdx)
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = uintFlag
+				t := row.GetTime(colIdx)
+				// Encoding timestamp need to consider timezone.
+				// If it's not in UTC, transform to UTC first.
+				if t.Type == mysql.TypeTimestamp && sc.TimeZone != time.UTC {
+					err = t.ConvertTimeZone(sc.TimeZone, time.UTC)
+					if err != nil {
+						return
+					}
+				}
+				var v uint64
+				v, err = t.ToPackedUint()
+				if err != nil {
+					return
+				}
+				b = (*[unsafe.Sizeof(v)]byte)(unsafe.Pointer(&v))[:]
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeDuration:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = durationFlag
+				// duration may have negative value, so we cannot use String to encode directly.
+				b = row.GetRaw(colIdx)
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeNewDecimal:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = decimalFlag
+				// If hash is true, we only consider the original value of this decimal and ignore it's precision.
+				dec := row.GetMyDecimal(colIdx)
+				b, err = dec.ToHashKey()
+				if err != nil {
+					return
+				}
+			}
+
+			err = writeColumn(w[i], buf, b)
+			if err != nil {
+				return
+			}
+		}
+	case mysql.TypeEnum:
+		for i := 0; i < rows; i++ {
+			row = chk.GetRow(i)
+			if row.IsNull(colIdx) {
+				buf[0], b = NilFlag, nil
+				nullBitmap[i/8] |= 1 << uint(i%8)
+			} else {
+				buf[0] = uvarintFlag
+				v := uint64(row.GetEnum(colIdx).ToNumber())
+				b = (*[8]byte)(unsafe.Pointer(&v))[:]
```

Comment:
We'd prefer not to use a magic digit here. For performance consideration, you can define a constraint when initializing.
```go
const sizeUint64 = unsafe.Sizeof(uint64(0))
```
Same as all the `unsafe` usages in this file.

Reply:
have done